Intersect.Server/Entities/ProjectileSpawn.cs

```diff
@@ -66,80 +66,74 @@ public bool IsAtLocation(Guid mapId, int x, int y, int z)
-            {
-                scalingStat = (Enums.Stats) Parent.Item.ScalingStat;
+                targetPlayer = (Player) targetEntity;
             }
 
             if (targetEntity != null && targetEntity != Parent.Owner)
             {
-
                 // Have we collided with this entity before? If so, cancel out.
-                if (mEntitiesCollided.Contains(en.Id))
+                if (mEntitiesCollided.Contains(targetEntity.Id))
                 {
                     if (!Parent.Base.PierceTarget)
                     {
+                        if(targetPlayer != null)
+                        {
+                            if(targetPlayer.Map.ZoneType == Enums.MapZones.Safe ||
+                                Parent.Owner is Player plyr && plyr.InParty(targetPlayer))
+                            {
+                                return false;
+                            }
+                        }
+
                         return true;
                     }
                     else
                     {
                         return false;
                     }
                 }
-                mEntitiesCollided.Add(en.Id);
+                mEntitiesCollided.Add(targetEntity.Id);
 
-                if (targetEntity.GetType() == typeof(Player)) //Player
+                if (targetPlayer != null)
                 {
                     if (Parent.Owner != Parent.Target)
                     {
                         Parent.Owner.TryAttack(targetEntity, Parent.Base, Parent.Spell, Parent.Item, Dir);
-                        if (Dir <= 3 && Parent.Base.GrappleHook && !Parent.HasGrappled
-                        ) //Don't handle directional projectile grapplehooks
+
+                        if (Dir <= 3 && ShouldHook(targetEntity) && !Parent.HasGrappled)
                         {
-                            Parent.HasGrappled = true;
-                            Parent.Owner.Dir = Dir;
-                            new Dash(
-                                Parent.Owner, Distance, (byte) Parent.Owner.Dir, Parent.Base.IgnoreMapBlocks,
-                                Parent.Base.IgnoreActiveResources, Parent.Base.IgnoreExhaustedResources,
-                                Parent.Base.IgnoreZDimension
-                            );
+                            HookEntity();
                         }
 
                         if (!Parent.Base.PierceTarget)
                         {
+                            if (targetPlayer.Map.ZoneType == Enums.MapZones.Safe ||
+                                Parent.Owner is Player plyr && plyr.InParty(targetPlayer))
+                            {
+                                return false;
+                            }
+
                             return true;
                         }
                     }
                 }
-                else if (targetEntity.GetType() == typeof(Resource))
+                else if (targetEntity is Resource)
```

Comment:
`targetEntity is Resource targetResource` and then use `targetResource` instead of the re-casted value below.